scripts/hal_alignment/maf_to_fasta.py

```diff
@@ -0,0 +1,246 @@
+#!/usr/bin/env python3
+
+# See the NOTICE file distributed with this work for additional information
+# regarding copyright ownership.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Convert each block of a multiple alignment format (MAF) file to FASTA format.
+
+Metadata are stored in a JSON file corresponding to each FASTA file.
+
+Examples::
+    python maf_to_fasta.py --genomes-file genomes.txt input.maf output_dir/
+
+"""
+
+from argparse import ArgumentParser
+import json
+import os
+from pathlib import Path
+import re
+from typing import Dict, Iterable, Pattern, Tuple, Union
+
+from Bio.AlignIO.MafIO import MafIterator
+from Bio import SeqIO
+from Bio.SeqRecord import SeqRecord
+
+
+def compile_maf_src_regex(genome_names: Iterable[str]) -> Pattern[str]:
+    """Make a MAF src field regex for the given genome names.
+
+    In a UCSC MAF file, the src field can be of the form '<genome>.<seqid>',
+    which is useful for storing both the genome and sequence name; these can then be extracted by taking
+    the substrings before and after the dot character ('.'), respectively. However, this cannot be done
+    unambiguously if there is a dot in either the genome or sequence name. Taking the names of genomes
+    known to be in a MAF file, this function creates a string Pattern object that can be used to extract
+    the genome and sequence names from a MAF src field of the form '<genome>.<seqid>'.
```

Comment:
Would it be worth re-indenting the whole paragraph?

Reply:
Yes, I will do this.